File: client/src/api/members.ts
import { apiRequest } from "../services/api";

export interface MemberRole {
  id: string;
  name: string;
  color: string | null;
  position: number;
}

export interface Member {
  user_id: string;
  pubkey: string;
  display_name: string | null;
  avatar_hash: string | null;
  is_bot: boolean;
  joined_at: string;
  roles: MemberRole[];
}

export interface BotApproval {
  pubkey: string;
  approved_by: string | null;
  approved_at: string | null;
  revoked_at: string | null;
  note: string | null;
}

interface ListPendingBotsResponse {
  bots: BotApproval[];
}

export interface Ban {
  pubkey: string;
  banned_by: string;
  reason: string | null;
  banned_at: string;
}

export interface AllowlistEntry {
  pubkey: string;
  added_by: string;
  added_at: string;
}

interface JoinMemberResponse {
  member: Member;
}

interface ListMembersResponse {
  members: Member[];
}

interface ListBansResponse {
  bans: Ban[];
}

interface ListAllowlistResponse {
  entries: AllowlistEntry[];
}

export async function joinServer(baseUrl: string, token: string): Promise<Member> {
  const response = await apiRequest<JoinMemberResponse>(baseUrl, "/api/v1/members/join", {
    method: "POST",
    token,
  });
  return response.member;
}

export async function leaveServer(baseUrl: string, token: string): Promise<void> {
  await apiRequest<void>(baseUrl, "/api/v1/members/me", { method: "DELETE", token });
}

export async function listMembers(baseUrl: string, token: string): Promise<Member[]> {
  const response = await apiRequest<ListMembersResponse>(baseUrl, "/api/v1/members", { token });
  return response.members;
}

export async function kickMember(baseUrl: string, token: string, pubkey: string): Promise<void> {
  await apiRequest<void>(
    baseUrl,
    `/api/v1/members/${encodeURIComponent(pubkey)}/kick`,
    { method: "POST", token },
  );
}

export async function banMember(
  baseUrl: string,
  token: string,
  pubkey: string,
  reason?: string,
): Promise<Ban> {
  return apiRequest<Ban>(
    baseUrl,
    `/api/v1/members/${encodeURIComponent(pubkey)}/ban`,
    {
      method: "POST",
      token,
      body: JSON.stringify({ reason }),
    },
  );
}

export async function listBans(baseUrl: string, token: string): Promise<Ban[]> {
  const response = await apiRequest<ListBansResponse>(baseUrl, "/api/v1/bans", { token });
  return response.bans;
}

export async function unbanMember(baseUrl: string, token: string, pubkey: string): Promise<void> {
  await apiRequest<void>(baseUrl, `/api/v1/bans/${encodeURIComponent(pubkey)}`, {
    method: "DELETE",
    token,
  });
}

export async function listAllowlist(baseUrl: string, token: string): Promise<AllowlistEntry[]> {
  const response = await apiRequest<ListAllowlistResponse>(baseUrl, "/api/v1/allowlist", { token });
  return response.entries;
}

export async function addAllowlist(baseUrl: string, token: string, pubkey: string): Promise<AllowlistEntry> {
  return apiRequest<AllowlistEntry>(baseUrl, "/api/v1/allowlist", {
    method: "POST",
    token,
    body: JSON.stringify({ pubkey }),
  });
}

export async function removeAllowlist(baseUrl: string, token: string, pubkey: string): Promise<void> {
  await apiRequest<void>(baseUrl, `/api/v1/allowlist/${encodeURIComponent(pubkey)}`, {
    method: "DELETE",
    token,
  });
}

export async function assignRole(
  baseUrl: string,
  token: string,
  pubkey: string,
  roleId: string,
): Promise<void> {
  await apiRequest<void>(
    baseUrl,
    `/api/v1/members/${encodeURIComponent(pubkey)}/roles/${encodeURIComponent(roleId)}`,
    { method: "PUT", token },
  );
}

export async function removeRole(
  baseUrl: string,
  token: string,
  pubkey: string,
  roleId: string,
): Promise<void> {
  await apiRequest<void>(
    baseUrl,
    `/api/v1/members/${encodeURIComponent(pubkey)}/roles/${encodeURIComponent(roleId)}`,
    { method: "DELETE", token },
  );
}

export async function listPendingBots(baseUrl: string, token: string): Promise<BotApproval[]> {
  const response = await apiRequest<ListPendingBotsResponse>(
    baseUrl,
    "/api/v1/admin/bots/pending",
    { token },
  );
  return response.bots;
}

export async function approveBot(
  baseUrl: string,
  token: string,
  pubkey: string,
  note?: string,
): Promise<BotApproval> {
  return apiRequest<BotApproval>(
    baseUrl,
    `/api/v1/admin/bots/${encodeURIComponent(pubkey)}/approve`,
    { method: "POST", token, body: JSON.stringify({ note }) },
  );
}

export async function revokeBot(
  baseUrl: string,
  token: string,
  pubkey: string,
): Promise<BotApproval> {
  return apiRequest<BotApproval>(
    baseUrl,
    `/api/v1/admin/bots/${encodeURIComponent(pubkey)}/revoke`,
    { method: "POST", token },
  );
}

File: sdk/rust/src/rest/members.rs
use crate::error::Result;
use crate::rest::models::{
    AllowlistEntry, AllowlistRequest, ApproveBotRequest, Ban, BanRequest, BotApproval,
    JoinMemberResponse, ListAllowlistResponse, ListBansResponse, ListMembersResponse,
    ListPendingBotsResponse, Member,
};
use crate::rest::Client;

pub struct MembersApi<'c> {
    client: &'c Client,
}

impl<'c> MembersApi<'c> {
    pub(crate) fn new(client: &'c Client) -> Self {
        Self { client }
    }

    pub async fn list(&self) -> Result<Vec<Member>> {
        let resp: ListMembersResponse = self.client.get("/api/v1/members").await?;
        Ok(resp.members)
    }

    pub async fn get(&self, pubkey: &str) -> Result<Member> {
        self.client
            .get(&format!("/api/v1/members/{pubkey}"))
            .await
    }

    /// POST `/api/v1/members/join` — join the server as the authenticated user.
    pub async fn join(&self) -> Result<Member> {
        let resp: JoinMemberResponse = self.client.post_empty("/api/v1/members/join").await?;
        Ok(resp.member)
    }

    pub async fn leave(&self) -> Result<()> {
        self.client.delete("/api/v1/members/me").await
    }

    pub async fn kick(&self, pubkey: &str) -> Result<()> {
        self.client
            .send_empty(
                reqwest::Method::POST,
                &format!("/api/v1/members/{pubkey}/kick"),
            )
            .await
    }

    pub async fn ban(&self, pubkey: &str, req: &BanRequest) -> Result<Ban> {
        self.client
            .post(&format!("/api/v1/members/{pubkey}/ban"), req)
            .await
    }

    pub async fn list_bans(&self) -> Result<Vec<Ban>> {
        let resp: ListBansResponse = self.client.get("/api/v1/bans").await?;
        Ok(resp.bans)
    }

    pub async fn unban(&self, pubkey: &str) -> Result<()> {
        self.client
            .delete(&format!("/api/v1/bans/{pubkey}"))
            .await
    }

    pub async fn list_allowlist(&self) -> Result<Vec<AllowlistEntry>> {
        let resp: ListAllowlistResponse = self.client.get("/api/v1/allowlist").await?;
        Ok(resp.entries)
    }

    pub async fn add_allowlist(&self, pubkey: impl Into<String>) -> Result<AllowlistEntry> {
        let body = AllowlistRequest { pubkey: pubkey.into() };
        self.client.post("/api/v1/allowlist", &body).await
    }

    pub async fn remove_allowlist(&self, pubkey: &str) -> Result<()> {
        self.client
            .delete(&format!("/api/v1/allowlist/{pubkey}"))
            .await
    }

    pub async fn list_pending_bots(&self) -> Result<Vec<BotApproval>> {
        let resp: ListPendingBotsResponse = self.client.get("/api/v1/admin/bots/pending").await?;
        Ok(resp.bots)
    }

    pub async fn approve_bot(
        &self,
        pubkey: &str,
        req: &ApproveBotRequest,
    ) -> Result<BotApproval> {
        self.client
            .post(&format!("/api/v1/admin/bots/{pubkey}/approve"), req)
            .await
    }

    pub async fn revoke_bot(&self, pubkey: &str) -> Result<BotApproval> {
        self.client
            .post_empty(&format!("/api/v1/admin/bots/{pubkey}/revoke"))
            .await
    }
}
